File: server/carrier_canary.py
"""Probe public carrier front doors without using real shipment identifiers."""

from __future__ import annotations

import argparse
import urllib.error
import urllib.request
from collections.abc import Callable, Mapping
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass
from typing import Any
from urllib.parse import urlparse

from .api_contract import CARRIER_CAPABILITIES

DEFAULT_TIMEOUT = 15.0
DEFAULT_ATTEMPTS = 2
MAX_WORKERS = 6


class _NoRedirectHandler(urllib.request.HTTPRedirectHandler):
    """A redirect proves reachability; do not follow it into a slower challenge page."""

    def redirect_request(
        self,
        request: urllib.request.Request,
        file_pointer: Any,
        code: int,
        message: str,
        headers: Any,
        new_url: str,
    ) -> None:
        return None


@dataclass(frozen=True)
class CanaryTarget:
    carrier_id: str
    display_name: str
    url: str

    @property
    def hostname(self) -> str:
        return urlparse(self.url).hostname or "invalid-host"


@dataclass(frozen=True)
class CanaryResult:
    target: CanaryTarget
    status: int | None
    error: str | None = None

    @property
    def healthy(self) -> bool:
        return self.status is not None and self.status < 500


def automatic_canary_targets(
    definitions: Mapping[str, Mapping[str, Any]] = CARRIER_CAPABILITIES,
) -> tuple[CanaryTarget, ...]:
    targets = []
    for carrier_id, definition in definitions.items():
        tracking = definition.get("tracking")
        if not isinstance(tracking, Mapping) or tracking.get("mode") != "automatic":
            continue
        display_name = definition.get("displayName")
        url = definition.get("canaryUrl")
        if not isinstance(display_name, str) or not isinstance(url, str):
            raise ValueError(f"Automatic carrier {carrier_id} has no canary metadata")
        parsed = urlparse(url)
        if (
            parsed.scheme != "https"
            or not parsed.hostname
            or parsed.username
            or parsed.password
            or parsed.query
            or parsed.fragment
        ):
            raise ValueError(f"Automatic carrier {carrier_id} has an unsafe canary URL")
        targets.append(CanaryTarget(carrier_id, display_name, url))
    return tuple(targets)


def request_status(url: str, timeout: float) -> int:
    request = urllib.request.Request(
        url,
        headers={
            "Accept": "text/html,application/json;q=0.8,*/*;q=0.5",
            "User-Agent": "SwissDeliveryTracker-Canary/1.0",
        },
    )
    try:
        opener = urllib.request.build_opener(_NoRedirectHandler())
        with opener.open(request, timeout=timeout) as response:
            return int(response.status)
    except urllib.error.HTTPError as exc:
        return exc.code


def probe_target(
    target: CanaryTarget,
    timeout: float = DEFAULT_TIMEOUT,
    attempts: int = DEFAULT_ATTEMPTS,
    fetch_status: Callable[[str, float], int] = request_status,
) -> CanaryResult:
    if attempts < 1:
        raise ValueError("Canary attempts must be positive")
    last_status: int | None = None
    last_error: str | None = None
    for _attempt in range(attempts):
        try:
            last_status = fetch_status(target.url, timeout)
            last_error = None
            if last_status < 500:
                break
        except (OSError, TimeoutError, urllib.error.URLError) as exc:
            last_status = None
            last_error = exc.__class__.__name__
    return CanaryResult(target, last_status, last_error)


def run_canaries(
    targets: tuple[CanaryTarget, ...],
    timeout: float = DEFAULT_TIMEOUT,
    attempts: int = DEFAULT_ATTEMPTS,
    fetch_status: Callable[[str, float], int] = request_status,
) -> tuple[CanaryResult, ...]:
    def probe(target: CanaryTarget) -> CanaryResult:
        return probe_target(target, timeout, attempts, fetch_status)

    with ThreadPoolExecutor(max_workers=min(MAX_WORKERS, len(targets) or 1)) as pool:
        return tuple(pool.map(probe, targets))


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--timeout", type=float, default=DEFAULT_TIMEOUT)
    parser.add_argument("--attempts", type=int, default=DEFAULT_ATTEMPTS)
    args = parser.parse_args(argv)

    targets = automatic_canary_targets()
    results = run_canaries(targets, args.timeout, args.attempts)
    for result in results:
        if result.healthy:
            print(
                f"PASS {result.target.carrier_id} {result.target.hostname} "
                f"HTTP {result.status}"
            )
        elif result.status is not None:
            print(
                f"FAIL {result.target.carrier_id} {result.target.hostname} "
                f"HTTP {result.status}"
            )
        else:
            print(
                f"FAIL {result.target.carrier_id} {result.target.hostname} "
                f"{result.error or 'unreachable'}"
            )
    healthy = sum(result.healthy for result in results)
    print(f"{healthy}/{len(results)} automatic carrier front doors reachable")
    return 0 if healthy == len(results) else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: src/server/carrierCanary.ts
import { pathToFileURL } from 'node:url';
import contract from '../../contracts/openapi.json' with { type: 'json' };

const DEFAULT_TIMEOUT_MS = 15_000;
const DEFAULT_ATTEMPTS = 2;
const MAX_CONCURRENCY = 6;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

export interface CanaryTarget {
  carrierId: string;
  displayName: string;
  url: string;
}

export interface CanaryResult {
  target: CanaryTarget;
  status: number | null;
  error?: string;
}

export function canaryHealthy(result: CanaryResult): boolean {
  return result.status !== null && result.status < 500;
}

export function automaticCanaryTargets(
  definitions: unknown = contract['x-carriers'],
): CanaryTarget[] {
  if (!isRecord(definitions)) throw new TypeError('Carrier definitions must be an object');
  const targets: CanaryTarget[] = [];
  for (const [carrierId, value] of Object.entries(definitions)) {
    if (!isRecord(value) || !isRecord(value.tracking) || value.tracking.mode !== 'automatic') {
      continue;
    }
    if (typeof value.displayName !== 'string' || typeof value.canaryUrl !== 'string') {
      throw new TypeError(`Automatic carrier ${carrierId} has no canary metadata`);
    }
    let url: URL;
    try {
      url = new URL(value.canaryUrl);
    } catch (error) {
      throw new TypeError(`Automatic carrier ${carrierId} has an unsafe canary URL`, { cause: error });
    }
    if (url.protocol !== 'https:' || !url.hostname || url.username || url.password || url.search || url.hash) {
      throw new TypeError(`Automatic carrier ${carrierId} has an unsafe canary URL`);
    }
    targets.push({ carrierId, displayName: value.displayName, url: url.toString() });
  }
  return targets;
}

export async function requestCanaryStatus(url: string, timeoutMs: number): Promise<number> {
  const response = await fetch(url, {
    headers: {
      Accept: 'text/html,application/json;q=0.8,*/*;q=0.5',
      'User-Agent': 'SwissDeliveryTracker-Canary/1.0',
    },
    cache: 'no-store',
    redirect: 'manual',
    signal: AbortSignal.timeout(timeoutMs),
  });
  try {
    await response.body?.cancel();
  } catch {
    // The HTTP status is sufficient even if the probe body cannot be cancelled.
  }
  return response.status;
}

export async function probeCanaryTarget(
  target: CanaryTarget,
  options: {
    timeoutMs?: number;
    attempts?: number;
    fetchStatus?: (url: string, timeoutMs: number) => Promise<number>;
  } = {},
): Promise<CanaryResult> {
  const timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  const attempts = options.attempts ?? DEFAULT_ATTEMPTS;
  const fetchStatus = options.fetchStatus ?? requestCanaryStatus;
  if (!Number.isInteger(attempts) || attempts < 1) throw new TypeError('Canary attempts must be positive');
  if (!Number.isFinite(timeoutMs) || timeoutMs <= 0) throw new TypeError('Canary timeout must be positive');
  let status: number | null = null;
  let error: string | undefined;
  for (let attempt = 0; attempt < attempts; attempt += 1) {
    try {
      status = await fetchStatus(target.url, timeoutMs);
      error = undefined;
      if (status < 500) break;
    } catch (caught) {
      status = null;
      error = caught instanceof Error ? caught.name : typeof caught;
    }
  }
  return { target, status, ...(error ? { error } : {}) };
}

export async function runCanaries(
  targets: CanaryTarget[],
  options: Parameters<typeof probeCanaryTarget>[1] = {},
): Promise<CanaryResult[]> {
  const results: CanaryResult[] = new Array(targets.length);
  let cursor = 0;
  const runWorker = async () => {
    while (cursor < targets.length) {
      const index = cursor;
      cursor += 1;
      results[index] = await probeCanaryTarget(targets[index]!, options);
    }
  };
  await Promise.all(
    Array.from({ length: Math.min(MAX_CONCURRENCY, targets.length) }, runWorker),
  );
  return results;
}

function optionValue(argv: string[], name: string, fallback: number): number {
  const index = argv.indexOf(name);
  if (index < 0) return fallback;
  const value = Number(argv[index + 1]);
  if (!Number.isFinite(value) || value <= 0) throw new TypeError(`${name} must be positive`);
  return value;
}

export async function carrierCanaryMain(argv = process.argv.slice(2)): Promise<number> {
  const timeoutMs = optionValue(argv, '--timeout', DEFAULT_TIMEOUT_MS / 1_000) * 1_000;
  const attempts = optionValue(argv, '--attempts', DEFAULT_ATTEMPTS);
  const results = await runCanaries(automaticCanaryTargets(), { timeoutMs, attempts });
  for (const result of results) {
    const hostname = new URL(result.target.url).hostname;
    if (canaryHealthy(result)) {
      console.log(`PASS ${result.target.carrierId} ${hostname} HTTP ${result.status}`);
    } else if (result.status !== null) {
      console.log(`FAIL ${result.target.carrierId} ${hostname} HTTP ${result.status}`);
    } else {
      console.log(`FAIL ${result.target.carrierId} ${hostname} ${result.error ?? 'unreachable'}`);
    }
  }
  const healthy = results.filter(canaryHealthy).length;
  console.log(`${healthy}/${results.length} automatic carrier front doors reachable`);
  return healthy === results.length ? 0 : 1;
}

const executable = process.argv[1] ? pathToFileURL(process.argv[1]).href : '';
if (import.meta.url === executable) {
  process.exitCode = await carrierCanaryMain();
}
